From 6653bf1a1094335289524ca098359b6a0117f800 Mon Sep 17 00:00:00 2001
From: Kafva <kafva.one@gmail.com>
Date: Sat, 25 Oct 2025 18:20:05 +0200
Subject: [PATCH 1/1] =?UTF-8?q?Cycle=20between=20tabs=20with=20=C3=A4/?=
 =?UTF-8?q?=C3=B6?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

There is an API for regular plugins to switch tabs:

    chrome.tabs.update(newTab.id, { active: true });

However, "chrome" does not appear to be available in pdf.js when it is
running in Firefox. The workaround in this patch hardwires global
mappings for switching tabs and an additional key to disable the special
behaviour. Note that ä/ö input will still be sent to focused text input
fields.
---
 toolkit/modules/ShortcutUtils.sys.mjs | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/toolkit/modules/ShortcutUtils.sys.mjs b/toolkit/modules/ShortcutUtils.sys.mjs
index 3e796cb6586a..5473d0761e92 100644
--- a/toolkit/modules/ShortcutUtils.sys.mjs
+++ b/toolkit/modules/ShortcutUtils.sys.mjs
@@ -344,7 +344,32 @@ export var ShortcutUtils = {
       !event.shiftKey &&
       !event.ctrlKey;
 
+    // HACK: use `.time` preference to avoid multiple triggers from one keypress
+    let now = Math.floor(Date.now() / 1000);
+    let tabSwitchActive = Services.prefs.getBoolPref("browser.tab_switch.active", false);
+    let tabSwitchLastSet = Services.prefs.getIntPref("browser.tab_switch.time", 0);
+    
     switch (event.keyCode) {
+      case 219: // 'å'  
+        if (event.altKey && meaningfulMetaKey && now > (tabSwitchLastSet + 2)) {
+            Services.prefs.setIntPref("browser.tab_switch.time", now);
+            Services.prefs.setBoolPref("browser.tab_switch.active", !tabSwitchActive);
+            console.log(`browser.tab_switch.active = ${!tabSwitchActive}`)
+        }
+        return null;
+      case 246: // 'ö'
+      case 186:
+      case 59:
+        if (tabSwitchActive) {
+            return ShortcutUtils.PREVIOUS_TAB;
+        }
+        break;
+      case 228: // 'ä'
+      case 222:
+        if (tabSwitchActive) {
+            return ShortcutUtils.NEXT_TAB;
+        }
+        break;
       case event.DOM_VK_TAB:
         if (event.ctrlKey && !event.altKey && !meaningfulMetaKey) {
           return ShortcutUtils.CYCLE_TABS;
-- 
2.51.1

